networkx/algorithms/isomorphism/isomorph.py

```diff
@@ -17,47 +17,91 @@
 
     Notes
     -----
-    Checks for matching degree, triangle, and number of cliques sequences.
     The triangle sequence contains the number of triangles each node is part of.
     The clique sequence contains for each node the number of maximal cliques
     involving that node.
 
+    Some properties are faster to compute than others. And there are other
+    properties we could include and don't. But of the three properties listed here,
+    comparing the degree distributions is the fastest. The "triangles" property
+    is slower (and also a stricter version of "could") and the "maximal cliques"
+    property is slower still, but usually faster than doing a full isomorphism
+    check.
     """
 
     # Check global properties
     if G1.order() != G2.order():
         return False
 
-    # Check local properties
-    d1 = G1.degree()
-    t1 = nx.triangles(G1)
-    clqs_1 = list(nx.find_cliques(G1))
-    c1 = Counter(chain.from_iterable(nx.find_cliques(G1)))  # number of cliques
-    props1 = [[d, t1[v], c1[v]] for v, d in d1]
-    props1.sort()
-
-    d2 = G2.degree()
-    t2 = nx.triangles(G2)
-    clqs_2 = list(nx.find_cliques(G2))
-    c2 = Counter(chain.from_iterable(nx.find_cliques(G2)))  # number of cliques
-    props2 = [[d, t2[v], c2[v]] for v, d in d2]
-    props2.sort()
-
-    if props1 != props2:
-        return False
-
-    # OK...
+    properties_to_check = set(properties)
+    G1_props, G2_props = [], []
+
+    def _properties_consistent():
+        # Ravel the properties into a table with # nodes rows and # properties columns
+        G1_ptable = [tuple(p[n] for p in G1_props) for n in G1]
+        G2_ptable = [tuple(p[n] for p in G2_props) for n in G2]
+
+        if sorted(G1_ptable) != sorted(G2_ptable):
+            return False
+        return True
+
+    # The property table is built and checked as each individual property is
+    # added. The reason for this is the building/checking the property table
+    # is in general much faster than computing the properties, making it
+    # worthwhile to check multiple times to enable early termination when
+    # a subset of properties don't match
+
+    # Degree sequence
+    if "d" in properties_to_check:
+        G1_props.append(G1.degree())
+        G2_props.append(G2.degree())
+        if not _properties_consistent():
+            return False
+    # Sequence of triangles per node
+    if "t" in properties_to_check:
+        G1_props.append(nx.triangles(G1))
+        G2_props.append(nx.triangles(G2))
+        if not _properties_consistent():
+            return False
+    # Sequence of maximal cliques per node
+    if "c" in properties_to_check:
+        G1_props.append(Counter(itertools.chain.from_iterable(nx.find_cliques(G1))))
+        G2_props.append(Counter(itertools.chain.from_iterable(nx.find_cliques(G2))))
+        if not _properties_consistent():
+            return False
+
+    # All checked conditions passed
     return True
```

Comment:
What if no conditions were checked (so `len(G1_props) == 0`) other than checking number of nodes? Should we warn or raise?

Reply:
Good question - it's a new corner-case given the flexibility of the `properties` kwarg.

My initial reaction is not to do any input validation. In other words, if users pass something different than the supported characters then they get the results of the `len(G) == len(H)` check.

If we did want input validation, we could check that the `properties` are a valid subset of the supported characters. Something like

```python
if not set(properties).issubset("dtc"):
    raise ValueError("Properties must be a combination of the characters 'd', 't', and/or 'c'")
```

I have a slight preference for the former (i.e. the current status quo: garbage-in-garbage-out!) but don't feel strongly!